manifests/charts/ztunnel/templates/daemonset.yaml

```
@@ -51,11 +51,70 @@ spec:
             add:
             - NET_ADMIN
         command:
+ {{- if .Values.ebpfRedirect }}
           - sh
           - -c
           - |
             set -ex
+            # tproxy mark, it's only used here.
+            MARK=0x400/0xfff
+            ORG_SRC_RET_MARK=0x4d3/0xfff
+            POD_OUTBOUND=15001
+            POD_INBOUND=15008
+            POD_INBOUND_PLAINTEXT=15006
+
+            EBPF_MARK_INBOUND=5678
+            EBPF_MARK_OUTBOUND=8765
+
+            set +e # Only for delete, we don't care if this fails
+            ip route flush table 100
+            set -e
+
+            ip rule add priority 20000 fwmark $MARK lookup 100
+            ip rule add priority 20003 fwmark $ORG_SRC_RET_MARK lookup 100
+            ip route add local 0.0.0.0/0 dev lo table 100
 
+            set +e
+            num_legacy_lines=$( (iptables-legacy-save || true; ip6tables-legacy-save || true) 2>/dev/null | grep '^-' | wc -l)
+            if [ "${num_legacy_lines}" -ge 10 ]; then
+              mode=legacy
+            else
+              num_nft_lines=$( (timeout 5 sh -c "iptables-nft-save; ip6tables-nft-save" || true) 2>/dev/null | grep '^-' | wc -l)
+              if [ "${num_legacy_lines}" -ge "${num_nft_lines}" ]; then
+                mode=legacy
+              else
+                mode=nft
+              fi
+            fi
+            IPTABLES=iptables-legacy
+            if [ "${mode}" = "nft" ]; then
+              IPTABLES=iptables-nft
+            fi
+            set -e
+
+            $IPTABLES -t mangle -F PREROUTING
+
+            $IPTABLES-save | grep -v LOG | $IPTABLES-restore
```

```suggestion
            $IPTABLES-save | grep -v LOG | $IPTABLES-restore
```

(might be my lack of iptables  knowledge showing) what's the point of this, exactly?  Saving rules, then immediately restoring them?

---

Good catch! It's copied from the very beginning ones. It's initially a debug purpose to log the tracing.  Save-filter-restore operation is just to make sure only one LOG entry for each chain. Anyway, we should it drop it right now(#42896 ).